process/headerCheck/headerVersioningHandler.go

```diff
@@ -0,0 +1,205 @@
+package headerCheck
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"sort"
+
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/storage"
+)
+
+const wildcard = "*"
+const keySize = 4
+
+type headerVersioningHandler struct {
+	referenceChainID []byte
+	versions         []config.VersionByEpochs
+	defaultVersion   string
+	versionCache     storage.Cacher
+}
+
+// NewHeaderVersioningHandler returns a new instance of a structure capable of handling the versions of a header
+// based on the provided epoch
+func NewHeaderVersioningHandler(
+	referenceChainID []byte,
+	versionsByEpochs []config.VersionByEpochs,
+	defaultVersion string,
+	versionCache storage.Cacher,
+) (*headerVersioningHandler, error) {
+
+	if len(referenceChainID) == 0 {
+		return nil, ErrInvalidReferenceChainID
+	}
+	if check.IfNil(versionCache) {
+		return nil, fmt.Errorf("%w, in NewHeaderVersioningHandler", ErrNilCacher)
+	}
+
+	hvh := &headerVersioningHandler{
+		referenceChainID: referenceChainID,
+		defaultVersion:   defaultVersion,
+		versionCache:     versionCache,
+	}
+	var err error
+	hvh.versions, err = hvh.prepareVersions(versionsByEpochs)
+	if err != nil {
+		return nil, err
+	}
+
+	err = hvh.checkVersionLength([]byte(defaultVersion))
+	if err != nil {
+		return nil, err
+	}
+
+	return hvh, err
+}
+
+func (hvh *headerVersioningHandler) prepareVersions(versionsByEpochs []config.VersionByEpochs) ([]config.VersionByEpochs, error) {
+	sort.Slice(versionsByEpochs, func(i, j int) bool {
+		return versionsByEpochs[i].StartEpoch < versionsByEpochs[j].StartEpoch
+	})
+
+	currentEpoch := uint32(0)
+	for _, ver := range versionsByEpochs {
+		if ver.StartEpoch >= ver.EndEpoch {
+			return nil, fmt.Errorf("%w, StartEpoch is greater or equal to EndEpoch, version %s",
+				ErrInvalidVersionOnEpochValues, ver.Version)
+		}
+		if currentEpoch != ver.StartEpoch {
+			return nil, fmt.Errorf("%w for version %s",
+				ErrInvalidVersionFoundGaps, ver.Version)
+		}
+		if len(ver.Version) > core.MaxSoftwareVersionLengthInBytes {
```

Comment:
The version should be also test for ascending I think

Reply:
It's not clear how we will encode the version at this step. Nor it is required to know about how it will look.

Reply:
ok